### PR TITLE
Document securing GraphQL fields (including associations) with `ApiProperty`

### PR DESCRIPTION
**Pull request by DavidBennettUK**
Documents https://github.com/api-platform/core/pull/4143

### CODE REVIEW

**Comment by alanpoulain** [comment]
Thank you for working on this @DavidBennettUK!

Shouldn't we talk about the nullability of the field too?

**Comment by DavidBennettUK** [comment]
@alanpoulain good catch! I've added a paragraph on that.

File: core/graphql.md [hunk 1]
```diff
@@ -1132,6 +1132,110 @@ class Book
 }
 ```
 
+### Securing properties (including associations)
```
Comment (alanpoulain) [inline core/graphql.md:1135]:
```suggestion
### Securing Properties (Including Associations)
```

File: core/graphql.md [hunk 2]
```diff
@@ -1132,6 +1132,110 @@ class Book
...
+
+You may want to limit access to certain resource properties with a security expression. This can be done with the `ApiProperty` `security` attribute.
+
+Note: adding the `ApiProperty` `security` expression to a GraphQL property will automatically make the GraphQL property type nullable (if it wasn't already).
+This is because `null` is returned as the property value if access is denied via the `security` expression.
+
+In GraphQL, it's possible to expose associations - allowing nested querying.
+For example, associations can be made with Doctrine ORM's OneToMany, ManyToOne, ManyToMany, etc.
+
+It's important to note that the security defined on resource operations applies only to the exposed query/mutation endpoints (e.g. `Query.users`, `Mutation.updateUser`, etc.)
```
Comment (alanpoulain) [inline core/graphql.md:1145]:
```suggestion
It's important to note that the security defined on resource operations applies only to the exposed query/mutation endpoints (e.g. `Query.users`, `Mutation.updateUser`, etc.).
```

File: core/graphql.md [hunk 3]
```diff
@@ -1132,6 +1132,110 @@ class Book
...
 
+### Securing properties (including associations)
+
+You may want to limit access to certain resource properties with a security expression. This can be done with the `ApiProperty` `security` attribute.
+
+Note: adding the `ApiProperty` `security` expression to a GraphQL property will automatically make the GraphQL property type nullable (if it wasn't already).
+This is because `null` is returned as the property value if access is denied via the `security` expression.
+
+In GraphQL, it's possible to expose associations - allowing nested querying.
+For example, associations can be made with Doctrine ORM's OneToMany, ManyToOne, ManyToMany, etc.
```
Comment (alanpoulain) [inline core/graphql.md:1143]:
```suggestion
For example, associations can be made with Doctrine ORM's `OneToMany`, `ManyToOne`, `ManyToMany`, etc.
```

File: core/graphql.md [hunk 4]
```diff
@@ -1132,6 +1132,110 @@ class Book
...
+
+It's important to note that the security defined on resource operations applies only to the exposed query/mutation endpoints (e.g. `Query.users`, `Mutation.updateUser`, etc.)
+Resource operation security is defined via the `security` attribute for each operation defined in the `ApiResource` `graphql` attribute.
+This security is *not* applied to exposed associations.
+
+Associations can instead be secured with the `ApiProperty` `security` attribute. This provides the flexibility to have different security depending on where an association is exposed.
+
+To prevent traversal attacks, you should ensure that any exposed associations are secured appropriately.
+A traversal attack is where a user can gain unintended access to a resource by querying nested associations, gaining access to a resource that prevents direct access (via the query endpoint).
+For example, a user may be denied using `Query.getUser` to get a user, but is able to assess the user through an association on an object that they do have access to (e.g. `document.createdBy`.)
```
Comment (alanpoulain) [inline core/graphql.md:1153]:
```suggestion
For example, a user may be denied using `Query.getUser` to get a user, but is able to access the user through an association on an object that they do have access to (e.g. `document.createdBy`.)
```

File: core/graphql.md [hunk 5]
```diff
@@ -1132,6 +1132,110 @@ class Book
...
+    /**
+     * @ORM\ManyToMany(targetEntity=User::class, inversedBy="viewableDocuments")
+     */
+    #[ApiProperty(security: 'is_granted("VIEW", object)')]
+    private Collection $viewers;
+
+    /**
+     * @ORM\ManyToOne(targetEntity=User::class)
+     */
+     #[ApiProperty(security: 'is_granted("VIEW", object)')]
```
Comment (alanpoulain) [inline core/graphql.md:1226]:
```suggestion
    #[ApiProperty(security: 'is_granted("VIEW", object)')]
```

**Review by alanpoulain — APPROVED** [review]

**Comment by alanpoulain** [comment]
Thank you very much @DavidBennettUK. Nice documentation.

**Pull request merged**